### DISCUSSION TITLE
Porting a Pygame-style game

### DISCUSSION THREAD

**Discussion by quantumjim** [nir/jupylet#29] [Q&A] [answered]
Is there any direct way to port a game with a manually implemented game loop (as one would do in Pygame)?

**Comment by nir** [comment, marked as answer]
Hi quantumjim, not sure what you mean. can you elaborate?

**Reply by quantumjim** [reply]
A Pygame-style game, as I'm thinking of it here, is basically some

```
while True:
    # get player input

    # change your sprites based on input

    # update screen
```

So the game loop is something defined within the game itself. Whereas in jupylet its more like

```
@game.qisge.app.run_me_every(1/FPS)
def next_frame()
    # get player input

    # change your sprites based on input

    # update screen

app.run()
```

where the game loop is hidden away in `app.run()`.

So it seems like a pygame-like game would have to be rewritten in this form in order to be run with jupylet. Is that the case?

**Reply by nir** [reply]
Jupylet relies on Python's asyncio event loop:

https://github.com/nir/jupylet/blob/538bf843c216f8c541c2a5afcb9f17ca514fac31/jupylet/app.py#L333

The dependency runs deep into Jupylet's event scheduling mechanisms so it cannot be removed easily.

In believe an ordinary Python loop such as the one you suggested above will prevent asyncio's event loop from functioning, since it kind of "hogs" the Python interpreter.

**Reply by quantumjim** [reply]
Thanks for the reply. That's as I expected. But you seem to be able to squeeze a lot more out of Jupyter than me, so I wondered if there was a way around it.

**Reply by nir** [reply]
Actually, Jupyter runs on top of a Python asyncio event loop (or its equivalent) - this is how a game (like spaceship) can run with graphics and all while the Jupyter notebook remains responsive.
